packages/pymedphys_dicom/src/pymedphys_dicom/dicom/_level2/anonymise.py

```diff
@@ -373,39 +375,111 @@ def anonymise_directory(
         information.
     """
     dicom_filepaths = glob(dicom_dirpath + '/**/*.dcm', recursive=True)
+    failing_filepaths = []
+    errors = []
 
     for dicom_filepath in dicom_filepaths:
-        anonymise_file(
-            dicom_filepath,
-            delete_original_file=False,
-            anonymise_filename=anonymise_filenames,
-            replace_values=replace_values,
-            keywords_to_leave_unchanged=keywords_to_leave_unchanged,
-            delete_private_tags=delete_private_tags,
-            delete_unknown_tags=delete_unknown_tags)
+        try:
+            anonymise_file(
+                dicom_filepath,
+                delete_original_file=False,
+                anonymise_filename=anonymise_filenames,
+                replace_values=replace_values,
+                keywords_to_leave_unchanged=keywords_to_leave_unchanged,
+                delete_private_tags=delete_private_tags,
+                delete_unknown_tags=delete_unknown_tags)
+        except Exception as e:
+            failing_filepaths.append(dicom_filepath)
+            errors.append(e)
 
     # Separate loop provides the ability to raise Exceptions from the
     # unsuccessful deletion of the original DICOM files while preventing
     # these Exceptions from interrupting the batch anonymisation.
     if delete_original_files:
         for dicom_filepath in dicom_filepaths:
-            remove_file(dicom_filepath)
+            if not dicom_filepath in failing_filepaths:
+                remove_file(dicom_filepath)
```

Comment:
I suspect we potentially shouldn't delete anything if there are any errors.

Reply:
Wait, did I misinterpret what you were saying [here](https://github.com/pymedphys/pymedphys/pull/222#discussion_r279619915)? I thought you preferred this approach and so implemented it that way.

Happy to fail fast :)

Reply:
Yeah let's just fail fast.

Reply:
Is this what you meant? It *is* what I had originally...

It fails fast in the sense the if an Error is throw during `anonymise_file()`, the program should simply exit (and thus the file deletion loop not executed).

Reply:
Yeah, sorry about that. I think a solution to the problem that was running through my head was to keep the output directory separate to the input directory. Meaning a half run program that still fails fast wouldn't have the outputs mixed in with the inputs, so the code could be repeated and it wouldn't try to use outputs as inputs for the next iteration.

I think this would be solved by using an implementation of output directory however. So we can leave that for another branch.

Sorry for sending you all over the place with that comment.

Reply:
All good.

We could ignore files with the "_Anonymous" label?

Reply:
Interesting solution ... It worries me though, if we start to ignore files like that. If I run anonymise on a file, no matter what its name I do expect it to run anonymisation over it. If we start having certain cases where that doesn't happen it starts to get quite confusing.

I think for my personal use I'll follow the separation of concerns pattern by keeping input directory and output directory apart from each other. We can sort that out at some point.